### PR TITLE
Add metrics compliance matrix

### LINKED ISSUE #1953: Create metrics feature spec compliance matrix
**What are you trying to achieve?**

https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md has a "TBD" for metrics feature compliance.

We need to fill in this table. @ocelotl has prepared a very compelling draft of this table, here:
https://github.com/open-telemetry/opentelemetry-python/discussions/2122

### PR DESCRIPTION
**Pull request by ocelotl**
Fixes #1953

### CODE REVIEW

File: table_generator.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+features = [
```
Comment (tigrannajaryan) [inline table_generator.py:1]: Do we need this generator file here? Do we intend to regenerate the table? If there is such intent it is best to have the actual list of features here.

Also, I don't know if we want this standalone Python file or it needs to be a part of the https://github.com/open-telemetry/build-tools which contains the generator for semantic conventions.

File: spec-compliance-matrix.md [hunk 2]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| When an invalid name is specified a working meter implementation is returned as a fallback.                                                                                       |          |    |      |    |        |      |        |     |      |     |      |       |
+| The fallback meter name property keeps its original invalid value                                                                                                                 |          |    |      |    |        |      |        |     |      |     |      |       |
+| The fallback meter name property keeps its original invalid value                                                                                                                 |          |    |      |    |        |      |        |     |      |     |      |       |
+| Test The API provides a way to set and get a global default MeterProvider                                                                                                         |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:101]:
```suggestion
| The API provides a way to set and get a global default MeterProvider                                                                                                         |          |    |      |    |        |      |        |     |      |     |      |       |
```

File: spec-compliance-matrix.md [hunk 3]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------|----|------|----|--------|------|--------|-----|------|-----|------|-------|
+| The API provides a way to set and get a global default MeterProvider                                                                                                              |          |    |      |    |        |      |        |     |      |     |      |       |
+| It is possible to create any number of MeterProviders                                                                                                                             |          |    |      |    |        |      |        |     |      |     |      |       |
+| MeterProvider provides getMeter                                                                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:97]:
```suggestion
| MeterProvider provides a way to get Meter                                                                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
```

File: spec-compliance-matrix.md [hunk 4]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| `get_meter` accepts name, version and schema_url                                                                                                                                  |          |    |      |    |        |      |        |     |      |     |      |       |
+| When an invalid name is specified a working meter implementation is returned as a fallback.                                                                                       |          |    |      |    |        |      |        |     |      |     |      |       |
+| The fallback meter name property keeps its original invalid value.                                                                                                                |          |    |      |    |        |      |        |     |      |     |      |       |
+| A message is logged reporting The specified value for the fallback meter is invalid.                                                                                              |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:107]: I guess this is not a must have.

File: spec-compliance-matrix.md [hunk 5]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| The meter provides functions to create a new Histogram                                                                                                                            |          |    |      |    |        |      |        |     |      |     |      |       |
+| The meter provides functions to create a new Asynchronous Gauge                                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
+| The meter provides functions to create a new UpDownCounter                                                                                                                        |          |    |      |    |        |      |        |     |      |     |      |       |
+| The meter provides functions to create a new                                                                                                                                      |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:114]:
```suggestion
| The meter provides functions to create a new Asynchronous UpDownCounter                                                                                                                                     |          |    |      |    |        |      |        |     |      |     |      |       |
```

File: spec-compliance-matrix.md [hunk 6]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| The instrument has an optional unit of measure.                                                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
+| The instrument has an optional description.                                                                                                                                       |          |    |      |    |        |      |        |     |      |     |      |       |
+| The meter returns an error when multiple instruments are registered under the same Meter using the same name.                                                                     |          |    |      |    |        |      |        |     |      |     |      |       |
+| It is possible to register two instruments with The same name under different meters.                                                                                             |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:120]:
```suggestion
| It is possible to register two instruments with same name under different Meters.                                                                                             |          |    |      |    |        |      |        |     |      |     |      |       |
```

File: spec-compliance-matrix.md [hunk 7]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| Instrument names conform to The specified syntax.                                                                                                                                 |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:121]:
```suggestion
| Instrument names conform to the specified syntax.                                                                                                                                 |          |    |      |    |        |      |        |     |      |     |      |       |
```

File: spec-compliance-matrix.md [hunk 8]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| Instrument units conform to The specified syntax.                                                                                                                                 |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:122]:
```suggestion
| Instrument units conform to the specified syntax.                                                                                                                                 |          |    |      |    |        |      |        |     |      |     |      |       |
```

File: spec-compliance-matrix.md [hunk 9]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| Instrument descriptions conform to The specified syntax.                                                                                                                          |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:123]:
```suggestion
| Instrument descriptions conform to the specified syntax.                                                                                                                          |          |    |      |    |        |      |        |     |      |     |      |       |
```

File: spec-compliance-matrix.md [hunk 10]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| The Counter can be created with create_counter.                                                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
+| The API Counter is an abstract class.                                                                                                                                             |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:125]: This is implementation specific, needs to be removed.

File: spec-compliance-matrix.md [hunk 11]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| Instrument names conform to The specified syntax.                                                                                                                                 |          |    |      |    |        |      |        |     |      |     |      |       |
+| Instrument units conform to The specified syntax.                                                                                                                                 |          |    |      |    |        |      |        |     |      |     |      |       |
+| Instrument descriptions conform to The specified syntax.                                                                                                                          |          |    |      |    |        |      |        |     |      |     |      |       |
+| The Counter can be created with create_counter.                                                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:124]: This probably needs to be reworded?

File: spec-compliance-matrix.md [hunk 12]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| The API for creating a counter accepts the unit of the instrument.                                                                                                                |          |    |      |    |        |      |        |     |      |     |      |       |
+| The API for creating a counter accepts the description of the instrument.                                                                                                         |          |    |      |    |        |      |        |     |      |     |      |       |
+| The counter has an add method.                                                                                                                                                    |          |    |      |    |        |      |        |     |      |     |      |       |
+| The add method returns None.                                                                                                                                                      |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:130]: Needs to be reworded since `None` is specific to Python.

File: spec-compliance-matrix.md [hunk 13]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| The add method accepts only positive amounts.                                                                                                                                     |          |    |      |    |        |      |        |     |      |     |      |       |
+| The Counter can be created with create_asynchronous_counter.                                                                                                                      |          |    |      |    |        |      |        |     |      |     |      |       |
+| The API ObservableCounter is an abstract class.                                                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
+| The API for creating a asynchronous counter accepts the name of the instrument.                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:136]:
```suggestion
| The API for creating an asynchronous counter accepts the name of the instrument.                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
```

File: spec-compliance-matrix.md [hunk 14]
```diff
@@ -90,9 +90,135 @@ formats is required. Implementing more than one format is optional.
...
+| The add method accepts optional attributes.                                                                                                                                       |          |    |      |    |        |      |        |     |      |     |      |       |
+| The add method accepts the increment amount.                                                                                                                                      |          |    |      |    |        |      |        |     |      |     |      |       |
+| The add method accepts only positive amounts.                                                                                                                                     |          |    |      |    |        |      |        |     |      |     |      |       |
+| The callback function has a timeout.                                                                                                                                              |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (reyang) [inline spec-compliance-matrix.md:145]: This description is not self contained.

**Review by reyang — DISMISSED** [review]
Please find my comments.

File: table_generator.py (hunk 1 above)
Comment (ocelotl) [reply table_generator.py:1]: I have found it useful, but I'll remove it from the time being since it is outside of the scope of this PR.

File: spec-compliance-matrix.md (hunk 4 above)
Comment (ocelotl) [reply spec-compliance-matrix.md:107]: Removed

File: spec-compliance-matrix.md (hunk 14 above)
Comment (ocelotl) [reply spec-compliance-matrix.md:144]: Adding a better description that refers to an asynchronous instrument

File: spec-compliance-matrix.md (hunk 12 above)
Comment (ocelotl) [reply spec-compliance-matrix.md:129]: Reworded it.

File: spec-compliance-matrix.md (hunk 11 above)
Comment (ocelotl) [reply spec-compliance-matrix.md:123]: Reworded it

File: spec-compliance-matrix.md (hunk 10 above)
Comment (ocelotl) [reply spec-compliance-matrix.md:124]: Removed

File: spec-compliance-matrix.md [hunk 15]
```diff
@@ -90,9 +90,134 @@ formats is required. Implementing more than one format is optional.
 
 ## Metrics
 
-|Feature                                       |Optional|Go|Java|JS |Python|Ruby|Erlang|PHP|Rust|C++|.NET|Swift|
-|----------------------------------------------|--------|--|----|---|------|----|------|---|----|---|----|-----|
-|TBD|
+| Feature                                                                                                                                                                           | Optional | Go | Java | JS | Python | Ruby | Erlang | PHP | Rust | C++ | .NET | Swift |
+|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------|----|------|----|--------|------|--------|-----|------|-----|------|-------|
+| The API provides a way to set and get a global default MeterProvider                                                                                                              |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (jsuereth) [inline spec-compliance-matrix.md:95]: Should this match the `Traces` compliance matrix more?  In that one, this would be listed as `MetricProvider` with a link to the metric provider specification and then "kind of sub-bullets" underneath with requirements for methods.

File: spec-compliance-matrix.md [hunk 16]
```diff
@@ -90,9 +90,134 @@ formats is required. Implementing more than one format is optional.
...
+| The Push exporter provides a shutdown function.                                                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
+| The Push exporter shutdown function return Failure after being called once.                                                                                                       |          |    |      |    |        |      |        |     |      |     |      |       |
+| The Push exporter shutdown function do not block indefinitely.                                                                                                                    |          |    |      |    |        |      |        |     |      |     |      |       |
+| The SDK provides OTEL_METRICS_EXEMPLAR_FILTER.                                                                                                                                    |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (jsuereth) [inline spec-compliance-matrix.md:216]: Environment variables should be split into the environment variable section.

File: spec-compliance-matrix.md [hunk 17]
```diff
@@ -90,9 +90,134 @@ formats is required. Implementing more than one format is optional.
...
+| The implementation keeps the last seen measurement that falls within an Histogram bucket.                                                                                         |          |    |      |    |        |      |        |     |      |     |      |       |
+| The reservoir will accept measurements by using the equivalent of the specified naive algorithm.                                                                                  |          |    |      |    |        |      |        |     |      |     |      |       |
+| The metric exporter has access to the pre aggregated metrics data                                                                                                                 |          |    |      |    |        |      |        |     |      |     |      |       |
+| The SDK provides a Push and a Pull Metric Exporter.                                                                                                                               |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (jsuereth) [inline spec-compliance-matrix.md:204]: The phrasing for this is odd

File: spec-compliance-matrix.md [hunk 18]
```diff
@@ -90,9 +90,134 @@ formats is required. Implementing more than one format is optional.
...
+| The Push exporter provides an export function.                                                                                                                                    |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (jsuereth) [inline spec-compliance-matrix.md:205]: This likely needs to update the `Exporter` section and clarify that the interface there is for Traces, and then that this one is for metrics.

File: spec-compliance-matrix.md [hunk 19]
```diff
@@ -90,9 +90,134 @@ formats is required. Implementing more than one format is optional.
...
+| The API for creating a asynchronous counter accepts the description of the instrument.                                                                                            |          |    |      |    |        |      |        |     |      |     |      |       |
+| The API for creating a asynchronous counter accepts a callback.                                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
+| The asynchronous counter has an add method.                                                                                                                                       |          |    |      |    |        |      |        |     |      |     |      |       |
+| The add method returns None.                                                                                                                                                      |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (jsuereth) [inline spec-compliance-matrix.md:140]: This looks like a python-specific thing (`None`)  Can we reword this to be more language-neutral?

reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#add

File: spec-compliance-matrix.md [hunk 20]
```diff
@@ -90,9 +90,134 @@ formats is required. Implementing more than one format is optional.
...
+| The add method accepts optional attributes.                                                                                                                                       |          |    |      |    |        |      |        |     |      |     |      |       |
+| The add method accepts the increment amount.                                                                                                                                      |          |    |      |    |        |      |        |     |      |     |      |       |
+| The add method accepts only positive amounts.                                                                                                                                     |          |    |      |    |        |      |        |     |      |     |      |       |
+| The callback function has a timeout.                                                                                                                                              |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (jsuereth) [inline spec-compliance-matrix.md:144]: Where is this requirement documented?

**Review by jsuereth — COMMENTED** [review]
Still in the middle of a review, lots of comments here.

File: spec-compliance-matrix.md (hunk 20 above)
Comment (ocelotl) [reply spec-compliance-matrix.md:144]: [Here it is](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#asynchronous-counter-creation): _The callback function SHOULD NOT take indefinite amount of time._

Comment (jsuereth) [reply spec-compliance-matrix.md:144]: A it's a "SHOULD", and additionally that seems like the requirement is targeted at users who give the callback functions, vs. the API having a required timeout.

Two points:

1. For Spec compliance, it'd be nice if we first focus on "MUST" requirements.  Would be able to update this PR so that MUST vs. SHOULD is somehow denoted?
2. Timeout isn't the only way to ensure a function completes in a finite amount of time.  I also don't think having a timeout attached to every callback would make a great API nor is it common.   To the extent we can keep similar language to the Specification for these requirements, that'd be ideal.

File: spec-compliance-matrix.md [hunk 21]
```diff
@@ -90,9 +90,134 @@ formats is required. Implementing more than one format is optional.
...
+| The Push exporter ForceFlush can inform the caller wether it succeeded, failed or timed out.                                                                                      |          |    |      |    |        |      |        |     |      |     |      |       |
+| The Push exporter provides a ForceFlush function.                                                                                                                                 |          |    |      |    |        |      |        |     |      |     |      |       |
+| The Push exporter provides a shutdown function.                                                                                                                                   |          |    |      |    |        |      |        |     |      |     |      |       |
+| The Push exporter shutdown function return Failure after being called once.                                                                                                       |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (jsuereth) [inline spec-compliance-matrix.md:214]: Again, this will be a comment on the interface / restriction on implementations.  I'm not sure the SDK itself needs to enforce this.

File: spec-compliance-matrix.md [hunk 22]
```diff
@@ -90,9 +90,134 @@ formats is required. Implementing more than one format is optional.
...
+| The Push exporter shutdown function do not block indefinitely.                                                                                                                    |          |    |      |    |        |      |        |     |      |     |      |       |
+| The SDK provides OTEL_METRICS_EXEMPLAR_FILTER.                                                                                                                                    |          |    |      |    |        |      |        |     |      |     |      |       |
+| The default value for OTEL_METRICS_EXEMPLAR_FILTER is WITH_SAMPLED_TRACE.                                                                                                         |          |    |      |    |        |      |        |     |      |     |      |       |
+| The value of NONE for OTEL_METRICS_EXEMPLAR_FILTER causes no measurements to be eligble for exemplar sampling.                                                                    |          |    |      |    |        |      |        |     |      |     |      |       |
```
Comment (jsuereth) [inline spec-compliance-matrix.md:218]: These environment variable values have changed (see: https://github.com/open-telemetry/opentelemetry-specification/pull/1965)

File: spec-compliance-matrix.md [hunk 23]
```diff
@@ -90,9 +90,134 @@ formats is required. Implementing more than one format is optional.
...
+| If a Measurement is reported by a synchronous Instrument, Then a MeasurementProcessor has access to the Baggage, Context and Span associated with the Measurement.                |          |    |      |    |        |      |        |     |      |     |      |       |
+| The metric SDK provides a mechanism to sample Exemplars from measurements.                                                                                                        |          |    |      |    |        |      |        |     |      |     |      |       |
+| The SDK allows an exemplar to be disabled.                                                                                                                                        |          |    |      |    |        |      |        |     |      |     |      |       |
+| A disabled exemplar does not cause any overhead.                                                                                                                                  |          |    |      |    |        |      |        |     |      |     |      |       |
```
[the remark this answers is not in the document]
Comment (ocelotl) [reply spec-compliance-matrix.md:178]: How do you call out a should requirement? Do you mark it in the optional column?

**Review by jmacd — APPROVED** [review]
Let's merge and iterate on this. I think we should find a way to incorporate SHOULD and MUST, but this is a good listing of features out of the current draft specification. Thanks @ocelotl!

File: spec-compliance-matrix.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (jmacd) [reply spec-compliance-matrix.md:178]: Let's write "A disabled exemplar SHOULD not cause any overhead." for now?

**Comment by ocelotl** [comment]
After reviewing feedback, I feel like a different approach will be needed for this PR, marking it as draft in order not to block progress on the metrics project, thanks for the reviews @tigrannajaryan, @jsuereth, @jmacd, @reyang. :v:
